src/Microsoft.Extensions.ML/ServiceCollectionExtensions.cs

```diff
@@ -19,27 +20,50 @@ public static class ServiceCollectionExtensions
         /// <param name="services">
         /// The <see cref="IServiceCollection "/> to add services to.
         /// </param>
+        /// <param name="implementationFactory">
+        /// The factory that creates the service.
+        /// </param>
         /// <returns>
         /// The <see cref="PredictionEnginePoolBuilder{TData, TPrediction}"/> that was added to the collection.
         /// </returns>
         public static PredictionEnginePoolBuilder<TData, TPrediction> AddPredictionEnginePool<TData, TPrediction>(
-            this IServiceCollection services)
+            this IServiceCollection services,
+            Func<IServiceProvider, PredictionEnginePool<TData, TPrediction>> implementationFactory)
             where TData : class
             where TPrediction : class, new()
         {
-            services.AddPredictionEngineServices<TData, TPrediction>();
+            services
+                .AddPrerequisiteServices()
+                .AddSingleton<PredictionEnginePool<TData, TPrediction>, PredictionEnginePool<TData, TPrediction>>(implementationFactory);
             return new PredictionEnginePoolBuilder<TData, TPrediction>(services);
         }
 
-        internal static IServiceCollection AddPredictionEngineServices<TData, TPrediction>(
+        /// <summary>
+        /// Adds a <see cref="PredictionEnginePoolBuilder{TData, TPrediction}"/> to the service collection.
+        /// </summary>
+        /// <param name="services">
+        /// The <see cref="IServiceCollection "/> to add services to.
+        /// </param>
+        /// <returns>
+        /// The <see cref="PredictionEnginePoolBuilder{TData, TPrediction}"/> that was added to the collection.
+        /// </returns>
+        public static PredictionEnginePoolBuilder<TData, TPrediction> AddPredictionEnginePool<TData, TPrediction>(
             this IServiceCollection services)
             where TData : class
             where TPrediction : class, new()
         {
-            services.AddLogging();
-            services.AddOptions();
-            services.TryAddEnumerable(ServiceDescriptor.Singleton<IPostConfigureOptions<MLOptions>, PostMLContextOptionsConfiguration>());
-            services.AddSingleton<PredictionEnginePool<TData, TPrediction>, PredictionEnginePool<TData, TPrediction>>();
+            services
+                .AddPrerequisiteServices()
+                .AddSingleton<PredictionEnginePool<TData, TPrediction>, PredictionEnginePool<TData, TPrediction>>();
+            return new PredictionEnginePoolBuilder<TData, TPrediction>(services);
+        }
+
+        private static IServiceCollection AddPrerequisiteServices(this IServiceCollection services)
```

Comment:
@eerhardt I'm considering making this method public and perhaps renaming it to something like `AddPredictionEnginePoolServices` or `AddPredictionEnginePoolPrerequisites` or `AddPredictionEnginePoolPrerequisiteServices` (or whatever you think is appropriate) because I'm in a situation where I need finer control on the registration.

Essentially, I want to create two `PredictionEnginePool`s and each one is using a custom `ModelLoader` as per the [SO question](https://stackoverflow.com/a/57698799/197591) this PR is based on.  However, the custom `ModelLoader.GetModel` method is costly as it performs a full train of the model.  This method gets called when the `PredictionEnginePool` is resolved during constructor injection.

So, I basically have a controller that injects multiple services as a collection:
```
public FooController(IEnumerable<IFooService> fooServices)
```

Each implementation of `IFooService` injects a `PredictionEnginePool`:
```
public BarService(PredictionEnginePool<Foo, Bar> predictionEnginePool)

public BazService(PredictionEnginePool<Foo, Baz> predictionEnginePool)
```

Of course the problem with this is that the models get trained as soon as the controller is created as everything is instantiated at that point.  However, I want to make the `PredictionEnginePool` services lazy so that they are only resolved when I first call a method on them.

I found a solution [here](https://stackoverflow.com/a/45775657/197591) to do that.  This is the code:
```
class LazyService<T> : Lazy<T> where T : class
{
    public LazyService(IServiceProvider serviceProvider) : base(serviceProvider.GetRequiredService<T>) {}
}
```

This means my service DI constructors become this:
```
public BarService(LazyService<PredictionEnginePool<Foo, Bar>> lazyPredictionEnginePool)

public BazService(LazyService<PredictionEnginePool<Foo, Baz>> lazyPredictionEnginePool)
```

And I just add this to my `Startup.ConfigureServices`:
```
services.AddPredictionEnginePool<Foo, Bar>();
services.AddPredictionEnginePool<Foo, Baz>();
services.AddSingleton<LazyService<PredictionEnginePool<Foo, Bar>>>();
services.AddSingleton<LazyService<PredictionEnginePool<Foo, Baz>>>();
```

However, this means I now have unnecessary registrations of `PredictionEnginePool<Foo, Bar>` and `PredictionEnginePool<Foo, Baz>` caused by the convenience method of `AddPredictionEnginePool` which I must call for all the other prerequisites, unless I call them myself, but this would then bloat my code a little.

By making the "prerequisites" method public, I can do the following instead which would not result in any unnecessary additional registrations, or indeed having all those prerequisites being called twice (!):
```
services.AddPredictionEnginePoolServices();
services.AddSingleton<LazyService<PredictionEnginePool<Foo, Bar>>>();
services.AddSingleton<LazyService<PredictionEnginePool<Foo, Baz>>>();
```

Let me know your thoughts on this.  It's a minor change and I think would be very helpful for users wanting a little more control on the registration of the `PredictionEnginePool`, but it would show as an additional exposed method on `services`. Perhaps then it could be in another extension namespace, or perhaps it's not a big issue at all.

Reply:
I think I'd need to see the full code to get my head around this scenario. Is it possible to share the full thing?

Reply:
I've created a temporary repo [here](https://github.com/NeoXtreem/prediction-engine-pool-sample-app) which is a stripped down version of my application to show just the relevant parts. I've left in code where necessary to show intent and to add context such as the custom `ModelLoader`, `IModelService` and `IPredictionService` concrete classes, etc. I've put a `TODO` comment in `Startup` to indicate where the effect of my proposed change would be.

Reply:
Thanks for taking the time to explain it, @NeoXtreem. I now understand why you want this. I think it makes sense to expose it. I'll comment on the new code inline.

Reply:
> I think this change looks good. Thanks for the contribution @NeoXtreem!
> 
> Once you fix up the doc comments, and the CI is green, I can merge.

@eerhardt I've updated the code to be more DRY by putting the common functionality of the two `AddPredictionEnginePool` methods into a private method called by each.  They each pass a callback delegate to do the specific thing which is to call the appropriate overload of `AddSingleton`.

Reply:
@eerhardt I've updated the code to be more DRY by putting the common functionality of the two AddPredictionEnginePool methods into a private method called by each. They each pass a callback delegate to do the specific thing which is to call the appropriate overload of AddSingleton.

Reply:
@eerhardt I've updated the code to be more DRY by putting the common functionality of the two `AddPredictionEnginePool` methods into a private method called by each. They each pass a callback delegate to do the specific thing which is to call the appropriate overload of `AddSingleton`.